README
The "ehstr.h" header file defines:

* utob, for printing values as strings of "0"s and "1"s
   (stdlib.h "strtoul" can be used to round-trip)
* revstr, to reverse the contents of a string in place
   (converts big to little endian)
* decimal_to_hex, to convert an strings of decimal digits to hex
   (arbitrarily long strings, treats as unsigned)
* hex_to_decimal, converts hex back to decimal
   (also arbitrarily long strings, also treats as unsigned)
* trimstr, to trim leading and trailing whitespace from a string
   (in place)
* strnlen, for environments which only have strlen


Packaging
---------
autoreconf -iv &&
 ./configure &&
 make &&
 make distcheck &&
 echo "Success."


License
-------
GNU Lesser General Public License (LGPL), version 3 or later.
See COPYING.GPL3 and COPYING.LGPL3 for details.


NEWS
libehstr NEWS
* Release version 1.0.1 (2016-05-21) [stable]
  * fixes unsafe array read

* Initial Release version 1.0.0 (2016-05-16) [stable]


tests/test-btou.c
#include "../src/ehstr.h"	/* nothing */

#include <stdio.h>		/* fprintf */
#include <stdlib.h>		/* strtol strtoul */
#include <echeck.h>		/* check_long check_unsigned_long */

int main(void)
{
	unsigned long result;
	int failures;
	int base2;

	failures = 0;

	base2 = 2;
	result = strtol("0101", NULL, base2);
	failures += check_long(result, 5);

	result = strtoul("101", NULL, base2);
	failures += check_unsigned_long(result, 5);

	if (failures) {
		fprintf(stderr, "%d failures in %s\n", failures, __FILE__);
	}

	return (failures > 127 || failures < -128) ? 127 : failures;
}


tests/test-utob.c
#include <stdio.h>
#include <string.h>
#include <echeck.h>
#include <values.h>		/* LONGBITS */

#include "../src/ehstr.h"

int main(void)
{
	int failures;
	char buf[LONGBITS + 1];
	size_t bits;

	failures = 0;

	bits = 4;
	utob(buf, LONGBITS + 1, 5, bits);
	failures += check_str("0101", buf);

	bits = 3;
	utob(buf, LONGBITS + 1, 5, bits);
	failures += check_str("101", buf);

	bits = 8;
	utob(buf, LONGBITS + 1, -3, bits);
	failures += check_str("11111101", buf);

	if (failures) {
		fprintf(stderr, "%d failures in %s\n", failures, __FILE__);
	}

	return (failures > 127 || failures < -128) ? 127 : failures;
}


tests/test-ehstr.c
#include <stdio.h>
#include <string.h>
#include <echeck.h>
#include <values.h>		/* LONGBITS */

#include "../src/ehstr.h"

int test_revstr(void)
{
	int failures;
	char expected[10], reverse[10];

	failures = 0;
	strncpy(expected, "edcba", 10);
	strncpy(reverse, "abcde", 10);

	revstr(reverse, 10);

	failures += check_str(reverse, expected);
	if (failures) {
		fprintf(stderr, "%d failures in check_revstr\n", failures);
	}

	return failures;
}

int test_trimstr(void)
{
	int failures;
	char expected[20], actual[20];

	failures = 0;

	strncpy(expected, "foo AND bar -> baz", 20);
	strncpy(actual, "foo AND bar -> baz\n", 20);
	trimstr(actual, 20);
	failures += check_str(actual, expected);

	strncpy(expected, "foo AND bar", 20);
	strncpy(actual, " \tfoo AND bar \n", 20);
	trimstr(actual, 20);
	failures += check_str(actual, expected);

	if (failures) {
		fprintf(stderr, "%d failures in check_trimstr\n", failures);
	}

	return failures;
}

int test_strnlen()
{
	int failures = 0;
	char *foo = "foo";

	failures += check_unsigned_int(3, strnlen(foo, 100));
	failures += check_unsigned_int(2, strnlen(foo, 2));

	if (failures) {
		fprintf(stderr, "%d failures in check_strnlen\n", failures);
	}
	return failures;
}

int test_utob(void)
{
	int failures;
	char buf[LONGBITS + 1];
	size_t bits;

	failures = 0;

	bits = 4;
	utob(buf, LONGBITS + 1, 5, bits);
	failures += check_str("0101", buf);

	bits = 3;
	utob(buf, LONGBITS + 1, 5, bits);
	failures += check_str("101", buf);

	bits = 8;
	utob(buf, LONGBITS + 1, -3, bits);
	failures += check_str("11111101", buf);

	if (failures) {
		fprintf(stderr, "%d failures in check_utob\n", failures);
	}
	return failures;
}

int test_btou(void)
{
	unsigned long result;
	int failures;
	int base2;

	failures = 0;

	base2 = 2;
	result = strtol("0101", NULL, base2);
	failures += check_long(result, 5);

	result = strtoul("101", NULL, base2);
	failures += check_unsigned_long(result, 5);

	if (failures) {
		fprintf(stderr, "%d failures in check_btou\n", failures);
	}
	return failures;
}

int test_decimal_to_hex()
{
	int failures;
	char buf[20];
	char *rv;
	const char *str;

	failures = 0;

	rv = decimal_to_hex("275", 3, buf, 20);

	if (rv == NULL) {
		fprintf(stderr, "NULL returned from decimal_to_hex\n");
		fprintf(stderr, "Aborting test\n");
		return (1 + failures);
	}

	failures += check_str(buf, "0x0113");

	rv = decimal_to_hex("65543", 10, buf, 20);

	if (rv == NULL) {
		fprintf(stderr, "NULL returned from decimal_to_hex\n");
		fprintf(stderr, "Aborting test\n");
		return (1 + failures);
	}

	failures += check_str(buf, "0x010007");

	str = "5088824049625";
	rv = decimal_to_hex(str, strlen(str), buf, 20);
	if (rv == NULL) {
		fprintf(stderr, "NULL returned from decimal_to_hex\n");
		fprintf(stderr, "Aborting test\n");
		return (1 + failures);
	}
	failures += check_str(buf, "0x04A0D58CBFD9");

	if (failures) {
		fprintf(stderr, "%d failures in test_decimal_to_hex\n",
			failures);
	}

	return failures;
}

int check_hex_to_decimal(char *in, char *expected)
{
	char buf[20];
	char *out;

	out = hex_to_decimal(in, strlen(in), buf, 20);
	if (out == NULL) {
		fprintf(stderr, "NULL returned from hex_to_decimal\n");
		return 1;
	}
	return check_str(out, expected);
}

int test_hex_to_decimal()
{
	int failures;

	failures = 0;

	failures += check_hex_to_decimal("0x00", "0");
	failures += check_hex_to_decimal("0x0113", "275");
	failures += check_hex_to_decimal("0x10007", "65543");

	if (failures) {
		fprintf(stderr, "%d failures in test_hex_to_decimal\n",
			failures);
	}

	return failures;
}

int test_decimal_to_hex_to_decimal_loop()
{
	int failures;
	int i;
	char buf[100];
	char hex[100];
	char dec[100];
	char *numv[] = { "0", "1", "10", "275", "65543", "17", "1025", "106" };
	int numc = 7;

	failures = 0;

	for (i = 0; i < numc; ++i) {
		decimal_to_hex(numv[i], strlen(numv[i]), hex, 100);
		hex_to_decimal(hex, 100, dec, 100);
		failures += check_str(dec, numv[i]);
	}

	/* lets just do a bunch */
	for (i = 1; i < 1000001; i += 25) {
		sprintf(buf, "%d", i);
		decimal_to_hex(buf, strlen(buf), hex, 100);
		hex_to_decimal(hex, 100, dec, 100);
		failures += check_str(dec, buf);
	}

	if (failures) {
		fprintf(stderr, "%d failures in test_hex_to_decimal\n",
			failures);
	}

	return failures;
}

/* int main(int argc, char *argv[]) */
int main(void)
{
	int failures = 0;

	failures += test_revstr();
	failures += test_trimstr();
	failures += test_strnlen();
	failures += test_utob();
	failures += test_btou();
	failures += test_decimal_to_hex();
	failures += test_hex_to_decimal();
	failures += test_decimal_to_hex_to_decimal_loop();

	if (failures) {
		fprintf(stderr, "%d failures in total\n", failures);
	}
	return failures;
}


estr.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <values.h>

#include "estr.h"

/*
  globals for controlling strictness and errors
*/
unsigned int ESTR_STRICT_UTOB = 0;
unsigned int ESTR_STRICT_BTOU = 0;
unsigned int ESTR_STRICT_STRNLEN = 0;
unsigned int ESTR_STRICT_REVSTR = 0;

/*
  unsigned to binary
  represents a binary value as a string of zeros and ones
*/
char *utob(char *buf, size_t buf_size, unsigned long val, size_t bits)
{
	size_t i, shift, str_pos;

	if (buf_size == 0) {
		fprintf(stderr, "buf_size of 0 is not valid\n");
		if (ESTR_STRICT_UTOB > 1) {
			exit(EXIT_FAILURE);
		} else if (ESTR_STRICT_UTOB > 0) {
			return NULL;
		}
		return buf;
	}
	if (bits == 0) {
		bits = LONGBITS;
	} else if (bits > LONGBITS) {
		fprintf(stderr, "%lu is greater than %lu\n",
			(unsigned long)bits, (unsigned long)LONGBITS);
		if (ESTR_STRICT_UTOB > 1) {
			exit(EXIT_FAILURE);
		}
		bits = LONGBITS;
	}

	str_pos = 0;
	for (i = bits; i > 0; i--) {
		if (str_pos >= (buf_size - 1)) {
			break;
		}
		shift = (i - 1);
		buf[str_pos++] = ((val >> shift) & 1) ? '1' : '0';
	}
	buf[str_pos] = '\0';

	return buf;
}

/*
  binary to unsigned
  convert a string of zeros and ones as a binary value
*/
unsigned long btou(const char *buf, size_t buf_size)
{
	size_t i, j, shift, len;
	unsigned long val, tmp;

	if (buf_size == 0) {
		fprintf(stderr, "buf_size of 0 is not valid\n");
		if (ESTR_STRICT_BTOU) {
			exit(EXIT_FAILURE);
		}
		return 0;
	}

	len = strnlen(buf, buf_size);

	val = buf[0] == '0' ? 0 : ULONG_MAX;

	for (i = 0, j = len; i < len; i++, j--) {
		shift = j - 1;
		if (buf[i] == '0') {
			tmp = ~(1L << shift);
			val &= tmp;
		} else if (buf[i] == '1') {
			tmp = 1L << shift;
			val |= tmp;
		} else {
			fprintf(stderr, "'%s' is not valid\n", buf);
			if (ESTR_STRICT_BTOU) {
				exit(EXIT_FAILURE);
			}
			return 0;
		}
	}
	return val;
}

#if _POSIX_C_SOURCE < 200809L
size_t strnlen(const char *str, size_t buf_size)
{
	size_t len = strlen(str);
	if (len > buf_size) {
		if (ESTR_STRICT_STRNLEN) {
			exit(EXIT_FAILURE);
		}
		return buf_size;
	}
	return len;
}
#endif

void revstr(char *str, size_t buf_size)
{
	size_t i, j, len;
	char swap;

	if (buf_size == 0) {
		if (ESTR_STRICT_REVSTR) {
			exit(EXIT_FAILURE);
		}
		return;
	}

	len = strnlen(str, buf_size);
	for (i = 0, j = len - 1; i < j; i++, j--) {
		swap = str[i];
		str[i] = str[j];
		str[j] = swap;
	}
}

unsigned int set_estr_strict_utob(unsigned int new_val)
{
	unsigned int previous = ESTR_STRICT_UTOB;
	ESTR_STRICT_UTOB = new_val;
	return previous;
}

unsigned int set_estr_strict_btou(unsigned int new_val)
{
	unsigned int previous = ESTR_STRICT_BTOU;
	ESTR_STRICT_BTOU = new_val;
	return previous;
}

unsigned int set_estr_strict_revstr(unsigned int new_val)
{
	unsigned int previous = ESTR_STRICT_REVSTR;
	ESTR_STRICT_REVSTR = new_val;
	return previous;
}

unsigned int set_estr_strict_strnlen(unsigned int new_val)
{
	unsigned int previous = ESTR_STRICT_STRNLEN;
	ESTR_STRICT_STRNLEN = new_val;
	return previous;
}


estr.h
#ifndef _ESTR_H_
#define _ESTR_H_

/*
  unsigned to big-endian binary
  represents a binary value as a string of zeros and ones
*/
char *utob(char *buf, size_t buf_szie, unsigned long val, size_t bits);

/*
  big-endian binary to unsigned
  convert a string of zeros and ones as a binary value
*/
unsigned long btou(const char *buf, size_t buf_size);

/*
  reverse a string in place
  e.g. to convert "0111" from big-endian to little-endian "1110"
*/
void revstr(char *str, size_t buf_size);

/*
  mostly for c89 compat
*/
#if _POSIX_C_SOURCE < 200809L
size_t strnlen(const char *str, size_t buf_size);
#endif

/*
  globals for controlling strictness and errors
  ... I guess globals are better than #define
*/
unsigned int set_estr_strict_utob(unsigned int new_val);
unsigned int set_estr_strict_btou(unsigned int new_val);
unsigned int set_estr_strict_revstr(unsigned int new_val);
unsigned int set_estr_strict_strnlen(unsigned int new_val);

#endif /* _ESTR_H_ */


src/ehstr.c
#include <values.h>		/* LONGBITS */
#include <ctype.h>		/* isspace */

#include "ehstr.h"

char *utob(char *buf, size_t buf_size, unsigned long val, size_t bits)
{
	size_t i, shift, str_pos;

	if (buf_size == 0) {
		return buf;
	}

	if (bits == 0 || bits > LONGBITS) {
		bits = LONGBITS;
	}

	str_pos = 0;
	for (i = bits; i > 0; i--) {
		if (str_pos >= (buf_size - 1)) {
			break;
		}
		shift = (i - 1);
		buf[str_pos++] = ((val >> shift) & 1) ? '1' : '0';
	}
	buf[str_pos] = '\0';

	return buf;
}

size_t ehstrnlen(const char *str, size_t buf_size)
{
	size_t i;

	for (i = 0; i < buf_size; ++i) {
		if (*(str + i) == '\0') {
			return i;
		}
	}

	return buf_size;
}

void trimstr(char *str, size_t buf_size)
{
	size_t i, j, offset, len, nonwhite;

	len = (str && buf_size) ? strnlen(str, buf_size) : 0;
	if (len == 0) {
		return;
	}

	nonwhite = 0;
	for (i = 0; i < len && isspace(str[i]); ++i) ;
	offset = i;

	if (offset) {
		for (i = 0, j = offset; j < len; ++i, ++j) {
			str[i] = str[j];
			if (isspace(str[i]) == 0) {
				nonwhite = i;
			}
		}
		if (nonwhite + 1 < len) {
			str[nonwhite + 1] = '\0';
		}
		return;
	}

	nonwhite = 0;
	for (i = len; i > 0 && nonwhite == 0; --i) {
		if (isspace(str[i - 1])) {
			str[i - 1] = '\0';
		} else {
			nonwhite = 1;
		}
	}
}

void revstr(char *str, size_t buf_size)
{
	size_t i, j, len;
	char swap;

	len = strnlen(str, buf_size);
	for (i = 0, j = len - 1; i < j; i++, j--) {
		swap = str[i];
		str[i] = str[j];
		str[j] = swap;
	}
}

static char nibble_to_hex(unsigned char nibble)
{
	if (nibble < 10) {
		return '0' + nibble;
	} else if (nibble < 16) {
		return 'A' + nibble - 10;
	} else {
		return '\0';
	}
}

char *decimal_to_hex(const char *dec_str, size_t dec_len, char *buf,
		     size_t buf_len)
{
	size_t i, j, k, hex_len;
	unsigned char *hex_buf;

	if (dec_str == 0 || buf == 0 || buf_len < 5) {
		return NULL;
	}

	buf[0] = '0';
	buf[1] = 'x';

	/* first operate with binary data, convert to ASCII later */
	/* start by adjusting the buf for our needs - needs to be unsigned */
	hex_buf = (unsigned char *)buf + 2;	/* skip past leading "0x" */
	hex_len = buf_len - 3;	/* and leave room for the NULL terminator */

	/* zero out the buffer */
	for (i = 0; i < hex_len; ++i) {
		hex_buf[i] = '\0';
	}

	for (i = 0; i < dec_len && dec_str[i] != '\0'; ++i) {
		if (dec_str[i] < '0' || dec_str[i] > '9') {
			buf[0] = '\0';
			return NULL;
		}
		/* we're doing another digit, multiply previous by 10 */
		for (j = 0; j < hex_len; ++j) {
			k = hex_len - 1 - j;
			hex_buf[k] *= 10;
		}

		hex_buf[hex_len - 1] += (dec_str[i] - '0');

		/* carry */
		for (j = 0; j < hex_len; ++j) {
			k = hex_len - 1 - j;
			if (hex_buf[k] >= 16) {
				hex_buf[k - 1] += (hex_buf[k] / 16);
				hex_buf[k] = (hex_buf[k] % 16);
			}
		}
	}

	/* convert to ASCII */
	for (j = 0; j < hex_len; ++j) {
		hex_buf[j] = nibble_to_hex(buf[2 + j]);
	}

	/* left shift away leading zeros */
	/* first find the index (j) of the first non-zero */
	for (j = 0; j < hex_len && hex_buf[j] == '0'; ++j) {
		;
	}

	/* but work on whole bytes */
	/* since j was incremented, whole bytes of '0' will be odd */
	if (j % 2 == 0) {
		j -= 1;
	}

	/* next, shift all the contents "j" places to the left */
	for (i = 0; i < (hex_len - j); ++i) {
		hex_buf[i] = hex_buf[i + j];
	}

	/* add a trailing NULL */
	buf[buf_len - 1 - j] = '\0';

	return buf;
}

char *hex_to_decimal(const char *hex, size_t hex_len, char *buf, size_t buf_len)
{
	size_t i, j, k, dec_len;
	unsigned char *dec_buf;
	char ascii_offset;
	unsigned char num_val;

	if (hex == 0 || buf == 0 || buf_len < 2 || buf_len < hex_len) {
		return NULL;
	}

	/* skip over leading '0x' in string */
	if (hex_len >= 2 && hex[0] == '0' && (hex[1] == 'x' || hex[1] == 'X')) {
		hex += 2;
		hex_len -= 2;
	}

	/* first operate with binary data, convert to ASCII later */
	/* start by adjusting the buf for our needs - needs to be unsigned */
	dec_buf = (unsigned char *)buf;
	dec_len = buf_len - 1;	/* leave room for the NULL terminator */

	/* zero out the buffer */
	for (i = 0; i < dec_len; ++i) {
		dec_buf[i] = 0;
	}

	for (i = 0; i < hex_len && hex[i] != 0; ++i) {
		ascii_offset = 0;
		if (hex[i] >= '0' && hex[i] <= '9') {
			ascii_offset = '0';
		} else if (hex[i] >= 'A' && hex[i] <= 'F') {
			ascii_offset = 'A';
		} else if (hex[i] >= 'a' && hex[i] <= 'f') {
			ascii_offset = 'a';
		}
		if (ascii_offset == 0) {
			dec_buf[0] = '\0';
			return NULL;
		}

		/* we're doing another digit, multiply previous by 16 */
		for (j = 0; j < dec_len; ++j) {
			k = dec_len - 1 - j;
			dec_buf[k] *= 16;
		}

		if (ascii_offset == '0') {
			num_val = (hex[i] - '0');
		} else {
			num_val = 10 + (hex[i] - ascii_offset);
		}
		dec_buf[dec_len - 1] += num_val;

		/* carry */
		for (j = 0; j < dec_len; ++j) {
			k = dec_len - 1 - j;
			if (dec_buf[k] >= 10) {
				dec_buf[k - 1] += (dec_buf[k] / 10);
				dec_buf[k] = (dec_buf[k] % 10);
			}
		}
	}

	/* convert to ASCII */
	for (j = 0; j < dec_len; ++j) {
		buf[j] = '0' + dec_buf[j];
	}

	/* left shift away leading zeros */
	/* first find the index (j) of the first non-zero */
	for (j = 0; j < dec_len && dec_buf[j] == '0'; ++j) {
		;
	}
	/* if everything is zero, include the last zero */
	if (j == dec_len) {
		--j;
	}
	/* next, shift all the contents "j" places to the left */
	for (i = 0; i < (dec_len - j); ++i) {
		dec_buf[i] = dec_buf[i + j];
	}

	/* add a trailing NULL */
	buf[buf_len - 1 - j] = '\0';

	return buf;
}


src/ehstr.h
#ifndef EHSTR_H
#define EHSTR_H

#include <stddef.h>		/* size_t */

/*
  unsigned to big-endian binary
  represents a binary value as a string of zeros and ones
  see also: strtoul(buf, NULL, 2);
*/
char *utob(char *buf, size_t buf_szie, unsigned long val, size_t bits);

/*
  reverse a string in place
  e.g. to convert "0111" from big-endian to little-endian "1110"
*/
void revstr(char *str, size_t buf_size);

/* trim leading and trailing whitespace from a string in place */
void trimstr(char *str, size_t len);

/* converts an arbitrarily long string of unsigned decimal digits into hex */
char *decimal_to_hex(const char *dec_str, size_t dec_len, char *buf,
		     size_t buf_len);

/* converts an arbitrarily long string of hex digits into unsigned decimal */
char *hex_to_decimal(const char *hex_str, size_t hex_len, char *buf,
		     size_t buf_len);

/*
  mostly for c89 compat

  return the length of str, but check only as many as buf_size;
  if a '\0' is not found in the first buf_size, return buf_size
*/
size_t ehstrnlen(const char *str, size_t buf_size);
#if _XOPEN_SOURCE < 700 && _POSIX_C_SOURCE < 200809L
#define strnlen ehstrnlen
#endif

#endif /* EHSTR_H */


test-estr.c
#include <stdio.h>
#include <string.h>
#include <echeck.h>
#include <values.h>		/* LONGBITS */

#include "estr.h"

int check_revstr(void)
{
	int failures;
	char expected[10], reverse[10];

	failures = 0;
	strncpy(expected, "edcba", 10);
	strncpy(reverse, "abcde", 10);

	revstr(reverse, 10);

	failures += check_str(reverse, expected);
	if (failures) {
		fprintf(stderr, "%d failures in check_revstr\n", failures);
	}

	return failures;
}

int check_strnlen()
{
	int failures = 0;
	char *foo = "foo";

	failures += check_unsigned_int(3, strnlen(foo, 100));
	failures += check_unsigned_int(2, strnlen(foo, 2));

	if (failures) {
		fprintf(stderr, "%d failures in check_strnlen\n", failures);
	}
	return failures;
}

int check_utob(void)
{
	int failures;
	char buf[LONGBITS + 1];
	size_t bits;

	failures = 0;

	bits = 4;
	utob(buf, LONGBITS + 1, 5, bits);
	failures += check_str("0101", buf);

	bits = 3;
	utob(buf, LONGBITS + 1, 5, bits);
	failures += check_str("101", buf);

	bits = 8;
	utob(buf, LONGBITS + 1, -3, bits);
	failures += check_str("11111101", buf);

	if (failures) {
		fprintf(stderr, "%d failures in check_utob\n", failures);
	}
	return failures;
}

int check_btou(void)
{
	unsigned long result;
	int failures;

	failures = 0;

	result = btou("0101", 4);
	failures += check_long(result, 5);

	result = btou("101", LONGBITS + 1);
	failures += check_long(result, -3);

	if (failures) {
		fprintf(stderr, "%d failures in check_btou\n", failures);
	}
	return failures;
}

/* int main(int argc, char *argv[]) */
int main(void)
{
	int failures = 0;

	failures += check_revstr();
	failures += check_strnlen();
	failures += check_utob();
	failures += check_btou();

	if (failures) {
		fprintf(stderr, "%d failures in total\n", failures);
	}
	return failures;
}


test-ehstr.c
#include <stdio.h>
#include <string.h>
#include <echeck.h>
#include <values.h>		/* LONGBITS */

#include "ehstr.h"

int test_revstr(void)
{
	int failures;
	char expected[10], reverse[10];

	failures = 0;
	strncpy(expected, "edcba", 10);
	strncpy(reverse, "abcde", 10);

	revstr(reverse, 10);

	failures += check_str(reverse, expected);
	if (failures) {
		fprintf(stderr, "%d failures in check_revstr\n", failures);
	}

	return failures;
}

int test_trimstr(void)
{
	int failures;
	char expected[20], actual[20];

	failures = 0;

	strncpy(expected, "foo AND bar -> baz", 20);
	strncpy(actual, "foo AND bar -> baz\n", 20);
	trimstr(actual, 20);
	failures += check_str(actual, expected);

	strncpy(expected, "foo AND bar", 20);
	strncpy(actual, " \tfoo AND bar \n", 20);
	trimstr(actual, 20);
	failures += check_str(actual, expected);

	if (failures) {
		fprintf(stderr, "%d failures in check_trimstr\n", failures);
	}

	return failures;
}

int test_strnlen()
{
	int failures = 0;
	char *foo = "foo";

	failures += check_unsigned_int(3, strnlen(foo, 100));
	failures += check_unsigned_int(2, strnlen(foo, 2));

	if (failures) {
		fprintf(stderr, "%d failures in check_strnlen\n", failures);
	}
	return failures;
}

int test_utob(void)
{
	int failures;
	char buf[LONGBITS + 1];
	size_t bits;

	failures = 0;

	bits = 4;
	utob(buf, LONGBITS + 1, 5, bits);
	failures += check_str("0101", buf);

	bits = 3;
	utob(buf, LONGBITS + 1, 5, bits);
	failures += check_str("101", buf);

	bits = 8;
	utob(buf, LONGBITS + 1, -3, bits);
	failures += check_str("11111101", buf);

	if (failures) {
		fprintf(stderr, "%d failures in check_utob\n", failures);
	}
	return failures;
}

int test_btou(void)
{
	unsigned long result;
	int failures;
	int base2;

	failures = 0;

	base2 = 2;
	result = strtol("0101", NULL, base2);
	failures += check_long(result, 5);

	result = strtoul("101", NULL, base2);
	failures += check_unsigned_long(result, 5);

	if (failures) {
		fprintf(stderr, "%d failures in check_btou\n", failures);
	}
	return failures;
}

int test_decimal_to_hex()
{
	int failures;
	char buf[20];
	char *rv;
	const char *str;

	failures = 0;

	rv = decimal_to_hex("275", 3, buf, 20);

	if (rv == NULL) {
		fprintf(stderr, "NULL returned from decimal_to_hex\n");
		fprintf(stderr, "Aborting test\n");
		return (1 + failures);
	}

	failures += check_str(buf, "0x0113");

	rv = decimal_to_hex("65543", 10, buf, 20);

	if (rv == NULL) {
		fprintf(stderr, "NULL returned from decimal_to_hex\n");
		fprintf(stderr, "Aborting test\n");
		return (1 + failures);
	}

	failures += check_str(buf, "0x010007");

	str = "5088824049625";
	rv = decimal_to_hex(str, strlen(str), buf, 20);
	if (rv == NULL) {
		fprintf(stderr, "NULL returned from decimal_to_hex\n");
		fprintf(stderr, "Aborting test\n");
		return (1 + failures);
	}
	failures += check_str(buf, "0x04A0D58CBFD9");

	if (failures) {
		fprintf(stderr, "%d failures in test_decimal_to_hex\n",
			failures);
	}

	return failures;
}

int check_hex_to_decimal(char *in, char *expected)
{
	char buf[20];
	char *out;

	out = hex_to_decimal(in, strlen(in), buf, 20);
	if (out == NULL) {
		fprintf(stderr, "NULL returned from hex_to_decimal\n");
		return 1;
	}
	return check_str(out, expected);
}

int test_hex_to_decimal()
{
	int failures;

	failures = 0;

	failures += check_hex_to_decimal("0x00", "0");
	failures += check_hex_to_decimal("0x0113", "275");
	failures += check_hex_to_decimal("0x10007", "65543");

	if (failures) {
		fprintf(stderr, "%d failures in test_hex_to_decimal\n",
			failures);
	}

	return failures;
}

int test_decimal_to_hex_to_decimal_loop()
{
	int failures;
	int i;
	char buf[100];
	char hex[100];
	char dec[100];
	char *numv[] = { "0", "1", "10", "275", "65543", "17", "1025", "106" };
	int numc = 7;

	failures = 0;

	for (i = 0; i < numc; ++i) {
		decimal_to_hex(numv[i], strlen(numv[i]), hex, 100);
		hex_to_decimal(hex, 100, dec, 100);
		failures += check_str(dec, numv[i]);
	}

	/* lets just do a bunch */
	for (i = 1; i < 1000001; i += 25) {
		sprintf(buf, "%d", i);
		decimal_to_hex(buf, strlen(buf), hex, 100);
		hex_to_decimal(hex, 100, dec, 100);
		failures += check_str(dec, buf);
	}

	if (failures) {
		fprintf(stderr, "%d failures in test_hex_to_decimal\n",
			failures);
	}

	return failures;
}

/* int main(int argc, char *argv[]) */
int main(void)
{
	int failures = 0;

	failures += test_revstr();
	failures += test_trimstr();
	failures += test_strnlen();
	failures += test_utob();
	failures += test_btou();
	failures += test_decimal_to_hex();
	failures += test_hex_to_decimal();
	failures += test_decimal_to_hex_to_decimal_loop();

	if (failures) {
		fprintf(stderr, "%d failures in total\n", failures);
	}
	return failures;
}
